support/src/run_every.h

```diff
@@ -52,7 +52,7 @@ namespace crimson {
 
     template<typename D>
     RunEvery(D                     _wait_period,
-	     std::function<void()> _body) :
+	     const std::function<void()>& _body) :
```

Comment:
Same here with a function pointer. As a general rule if you're using small/built-in types (ints, chars, iterators, etc.) usually it's best to pass them by const value if you in effect want to make a copy.

Reply:
I understand, but std::function is not the same as a function pointer as far as I know. They are similar indeed, but a pointer cannot hold data other than the address it points to. An std::function can copy the function which this const reference will avoid if its not necessary.

Reply:
std::function<> is a bit of a magical beast, as you're right, it's not the same as a function pointer. Instead, it's wrapping a Callable. I looked into this in more detail, and the answer is interesting: the code generated for a function pointer or reference-to-lambda is impressively concise. So the real question is "what happens when we copy a std::function that's pointing to some big Callable object" and the answer is... it will have to do the right thing and make a copy. Which could be expensive. 
So, one of two things needs to happen in that case: either the object needs to be moved, or the cost of that copy is something that's expected. So, you're right: if you want to avoid that happening at all, passing by const& looks like the way to go! Thanks for giving me the opportunity to look into it more.

Reply:
Anytime! And thanks for the explanation! I was going to do that myself since I really didn't recall the details.